samcli/local/docker/container.py

```diff
@@ -127,15 +125,47 @@ def create(self):
         if self.network_id == 'host':
             kwargs["network_mode"] = self.network_id
 
+        if not self.should_put_archive():
```

Comment:
I am not super crazy about having to do a check here and another check after the container is created (line 138). I have to do this because if we need to put the archive we shouldn't be telling the container to mount it on create. There has to be a better way to make this cleaner but not coming up with anything. @sanathkr or @TheSriram: Would love to hear if you have thoughts on how to clean this part up.

Reply:
copy is

```
self.docker_client.containers.create(self._image, **kwargs)
self._put_archives(docker_volumes, real_container)
```

mount is

```
kwargs["volumes"] = {to_posix_path(host_dir): mount for host_dir, mount in docker_volumes.items()}
self.docker_client.containers.create(self._image, **kwargs)
```

out of which below is common

```
self.docker_client.containers.create(self._image, **kwargs)
```

can we hide the order in which we call (mount,copy) through a wrapped decorator, where the decorator arguments determine which way to wrap the function? 

```
@files_relay(type={assigned based on some other logic eg: "copy" if go else "mount"}, volumes=docker_volumes)
def start_container(self._image, **kwargs):
   self.docker_client.containers.create(self._image, **kwargs)
```

```
def files_relay(type, volumes):
    def wrap(f):
        def wrapped_f(*args, **kwargs):
           if type == 'copy':
              real_container = f(*args,**kwargs)
              self._put_archives(volumes, real_container)
          else:
             kwargs["volumes"] = {to_posix_path(host_dir): mount for host_dir, mount in docker_volumes.items()}
             f(*args,**kwargs)
        return wrapped_f
    return wrap
```

then we just call 

```
start_container(self._image, **kwargs)
```

Notice how we are calling and changing the order of 'f' in the above, Does something like that work?